### ISSUE TITLE
ViewStart Script Ordering

### ISSUE THREAD

**Issue by Randy-Buchholz** [dotnet/aspnetcore#27812]
ViewStart runs before layout and page code. It is like like a "boot" for the process. Like other `.cshtml` files it can contain html/script data. This data gets merged into the final page.  Currently `<script>` tags in ViewStart are merged into the `body` section of the page _after_  script tags in the page. Since ViewStart is "first" and can contain common information, these scripts should be inserted before any layout or pages scripts.

### Use Case
Where ViewStart contains information used to boot a page in the browser.  
The script in ViewStart needs to complete before the script in the page runs. This can be done by creating a "Gate" using a `Promise` . The implementation of "Gate" is not shown, but it is awaitable on its `signal` property and blocks until `continue()` is called.

//_ViewStart.cshtml
```
<script>
    globalThis.bootWaiter = new Gate();
    // ... boot code
    bootWaiter.continue();
</script>
```

// Page.cshtml
```
<script>
    (async () => {
        await bootWaiter.signal;
        .// .. run scripts that depend on ViewStart
    })();
</script>
```

Currently the script in `ViewStart` ends up in the rendered page after the script in Page, and `bootWaiter` will be undefined.  

Workarounds are not as straight-forward as they could be. Even though ViewStart is processed first in the Razor phase, it (or its content) is not directly visible to Layouts or Pages. Options like `Section` aren't available until Layout. The lack of visibility into ViewStart is a limitation. There are options. One that works is an injected "Bag".

// ViewStart
```
@inject Bag bag
bag.AddBagItem("BootWaiter", new HtmlString("globalThis.bootWaiter = new Gate();"))
```

// Page/Layout
```
@inject Bag bag
<head>
    @bag.Content("BootWaiter")
</head>
```
The `Bag` allows sharing ViewStart data with Layout/Page processing. (Maybe a Bag should be built in.)

This also shows another limited aspect of script merging. All scripts are merged into `body`. Since ViewStart runs early there is a good chance some of these scripts really belong in `head`.

## Proposed Solution
Ensure that `<script>` content is merged into the final page before any `<script>` content in Layouts or Pages.

Option: Support Head Merging
// ViewStart
```
<head>
    <script></script>
</head>
```
Any scripts in a head group are merged into the page head.

**Comment by pranavkm** [comment]
@Randy-Buchholz a lot of the behavior in _ViewStart exists as back-compat with MVC5. Most markup changes that can be achieved using a _ViewStart can also be solved by a combination of `@section` or Layouts (or both). This makes it very unlikely for us to add additional features to this area. Just an FYI in case  you don't see any progress in this issue.

**Comment by Randy-Buchholz** [comment]
@pranavkm thanks for the info. I was using it less for basic markup and more as a setup for my layouts. I have different "base layouts" for different targets (mobile, desktop) and I was using _ViewStart to aggregate common info used by these. I can get the same thing done with `@inject`. I'm going to close this.
